### PR TITLE
screen: Use `legacysupport` to support OSX 10.6

### PR DESCRIPTION
**Pull request by TheRealKeto**
#### Descriptio

Fix `screen` on OSX 10.6 by using the `legacysupport` PG for `strndup`. Other patches were added to use appropriate functions and fix building on OSX 10.5 and lower.

###### Type(s)

- [x] bugfix

###### Tested on

macOS 10.7.5 11G63 x86_64
Xcode 4.6.3 4H1503

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -d install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by TheRealKeto** [comment]
@barracuda156 Would you mind testing this on OSX 10.6? I'm a bit skeptical about other possible undefined symbols that macports-legacy-support could provide.

**Review by dgilman — APPROVED** [review]

File: sysutils/screen/Portfile
Comment (barracuda156) [inline sysutils/screen/Portfile:11]: If it failed to build, there is no need to revbump it.

**Review by barracuda156 — CHANGES_REQUESTED** [review]

**Comment by barracuda156** [comment]
@TheRealKeto I tried to build this with `legacysupport` added, and it failed at linking on:
```
ccache /opt/local/bin/gcc-mp-14 -Wl,-headerpad_max_install_names -L/opt/local/lib -lMacportsLegacySupport -arch ppc -o screen screen.o acls.o ansi.o attacher.o backtick.o canvas.o comm.o display.o encoding.o fileio.o help.o input.o kmapdef.o layer.o layout.o list_display.o list_generic.o list_license.o list_window.o logfile.o mark.o misc.o process.o pty.o resize.o sched.o search.o socket.o telnet.o term.o termcap.o tty.o utmp.o viewport.o window.o winmsg.o winmsgbuf.o winmsgcond.o -lcurses  -lpam
Undefined symbols:
  "__vprocmgr_detach_from_console", referenced from:
      _main in screen.o
ld: symbol(s) not found
collect2: error: ld returned 1 exit status
make: *** [screen] Error 1
```
This says nothing to me, but likely something is missing here.

**Comment by barracuda156** [comment]
By the way, this is wrong in `screen.c`:
```
extern char **environ;
```
Should be fixed to use a proper Apple version of the definition.

**Comment by barracuda156** [comment]
This may be a conflict with system headers. `vproc_priv.h` apparently exists in macOS, for example used here: https://opensource.apple.com/source/shell_cmds/shell_cmds-149/nohup/nohup.c.auto.html
But `screen` bundles its own local `vproc_priv.h` (looks like a poor choice).

**Comment by barracuda156** [comment]
I also dunno how this gonna work on 10.5, where `__MAC_10_6` is probably undefined.
```
void _vproc_standby_begin(void)																__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
void _vproc_standby_end(void)																__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
size_t _vproc_standby_count(void)															__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
size_t _vproc_standby_timeout(void)															__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);

kern_return_t _vproc_transaction_count_for_pid(pid_t p, int32_t *count, bool *condemned)	__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
bool _vproc_pid_is_managed(pid_t p)															__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
void _vproc_transaction_try_exit(int status)												__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
void _vproc_transaction_begin(void)															__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
void _vproc_transaction_end(void)															__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
size_t _vproc_transaction_count(void)														__OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_NA);
```
(This is from the bundled copy of `vproc_priv.h`.)

**Comment by barracuda156** [comment]
Any idea where the symbol is supposed to come from? It is not there is `libSystem` in 10.5.8, for example:
```
36-25% nm -arch ppc -g /Volumes/Leo/usr/lib/libSystem.B.dylib | grep vprocmgr
000492ec T __vprocmgr_getsocket
0015c468 T __vprocmgr_init
0015c56c T __vprocmgr_log_drain
0015c524 T __vprocmgr_log_forward
00083e88 T __vprocmgr_move_subset_to_user
```

**Comment by barracuda156** [comment]
Oh well, this is MacPorts likely messing something up. I just noticed:
```
platform darwin {
    # The vproc_priv.h is the private header from launchd
    # We copied the header file from http://launchd.macosforge.org/trac/browser/trunk/launchd/src/vproc_priv.h, r23776
    distfiles-append    vproc_priv.h:launchd_priv_header
    post-extract {
        file copy ${distpath}/vproc_priv.h ${worksrcpath}
    }
}
```

**Comment by barracuda156** [comment]
@ryandesign @dgilman Could some comment on the issue please? From portfile it follows that the code is supposed to work on 10.5 (which implies it should build on 10.6 ppc either). Where is `__vprocmgr_detach_from_console` supposed to come from? Is it not present in libSystem in 10.5.8, AFAICT.

**Comment by TheRealKeto** [comment]
@barracuda156 After a little bit of digging through git blame (as well as a large portion "googling"), `vproc_priv.h` comes from `launchd`. I believe [this particular revision closely resembles the one dowloaded by MacPorts](https://github.com/apple-oss-distributions/launchd/blob/launchd-328/launchd/src/vproc_priv.h)

Unsure on how to proceed given the many revisions that the header has in conjuction with the many OSX 10.6 releases, but that's where the header comes from.

**Comment by barracuda156** [comment]
@TheRealKeto Could you pick the commit for `environ`? https://github.com/macports/macports-ports/commit/bcce82a242ce7321e76ff9a11713a7dc23938f3e
It is against your tree, so hopefully works.

Not sufficient to fix the above issue with 10.5 (and 10.6?), but still needed.

**Comment by barracuda156** [comment]
> @barracuda156 After a little bit of digging through git blame (as well as a large portion "googling"), `vproc_priv.h` comes from `launchd`. I believe [this particular revision closely resembles the one dowloaded by MacPorts](https://github.com/apple-oss-distributions/launchd/blob/launchd-328/launchd/src/vproc_priv.h)
> 
> Unsure on how to proceed given the many revisions that the header has in conjuction with the many OSX 10.6 releases, but that's where `__vprocmgr_detach_from_console` comes from.

Well, you do not have to be concerned about pre-release versions of 10.6, of course, however the port claims that this is supported on 10.5. I do not see how it is. I cannot check it now, but looks like it should fail to link there.

If so, then the patch must not be used on 10.5, just like it is not used on 10.4. I cannot test on 10.6 x86_64 right now, so cannot say if it works there or not.

**Comment by barracuda156** [comment]
@TheRealKeto If you have 10.6.8 at hand, could you please check if `__vprocmgr_detach_from_console` is present in `libSystem`?

**Comment by TheRealKeto** [comment]
@barracuda156 Unfortunately, I don't have 10.6.8. Let's wait for Ryan's response regarding the issue; thanks for helping me along the way :)

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
@barracuda156

zsh1

**Comment by barracuda156** [comment]
> @barracuda156
> 
> zsh1

@aeiouaeiouaeiouaeiouaeiouaeiou Thank you, then the remaining question is what to do about 10.5, which does not seem to have this.

**Comment by dgilman** [comment]
Many years ago I explicitly dropped 10.5 support from this package. Someone may have attempted to re-add it, I have no idea if that works. I am OK with marking this package as not working on 10.5, I do not value compatibility with these old systems.

if you absolutely want to get it working on 10.5 I think you need to redo the patches entirely or ship an older version of screen. Either way you are looking at a new subport or a second port file.

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
The fact that it doesn't build on 10.5 isn't a critical issue, so let's merge it.

**Review by aeiouaeiouaeiouaeiouaeiouaeiou — APPROVED** [review]

**Comment by barracuda156** [comment]
@dgilman @aeiouaeiouaeiouaeiouaeiouaeiou Does something stop us from _not_ applying a patch which breaks it for 10.5? After all, it is not the original source which is the problem, it is MacPorts breaks it by using a patch. It is already disabled on 10.4, everything that is needed is to use instead a condition for 10.4–10.5 (ignore 10.6 ppc, that I can add locally).

**Comment by dgilman** [comment]
So the plan is to drop both `patch-apple-*` patches for 10.4 and 10.5? Yes, that's totally doable. You'll lose the utmp accounting but that's not load bearing / I imagine next to nobody uses it anyway.

**Comment by barracuda156** [comment]
@dgilman I think it is an okay solution, especially given that it is already done for darwin8, so only the condition is to be modified. All the needed code is already there.

**Comment by TheRealKeto** [comment]
Any more changes before this is marked as ready to be merged?

**Comment by barracuda156** [comment]
> Any more changes before this is marked as ready to be merged?

@TheRealKeto If you could change `darwin 8` to Darwin < 10 (so that we do not break 10.5, as discussed above), that would be great. Other than that all good, I guess.

File: sysutils/screen/Portfile
Comment (barracuda156) [inline sysutils/screen/Portfile:82]: Sorry, I was replying from mobile earlier, it was difficult to type special characters.
It should be:
```
if {${os.platform} eq "darwin" && ${os.major} < 10} {
```

Comment (TheRealKeto) [reply sysutils/screen/Portfile]: No worries, I suspected that I'd done it wrong. Fixed in latest commit

**Review by barracuda156 — APPROVED** [review]
Thank you very much

**Review by dgilman — APPROVED** [review]

**Pull request merged**
